### DISCUSSION TITLE
Questions about PhysicsObject.gd

### DISCUSSION THREAD

**Discussion by TheDrewlander** [Techokami/SonicWorldsNext#130] [Q&A]
Hello. Bare with me since this is the first time I've actually posted a discussion on GitHub (or any forum to be honest). I just wanted to gain more insight as to how the physics work. You see, I'm also trying to program my own sonic physics platformer, but I am struggling to understand how exactly sonic/the player is able to orient/rotate around a curve seamlessly while still keeping the gravity pushing downwards whenever they lose speed/fall off a curve or start jumping.

I took a look at the PhysicsObject.gd script, but I'm really confused what all the variables and numbers are doing and why it's done this way. Specifically, I'm trying to understand the _physics_process loop. Why does it need to cut down the movement with moveRemaining and the while loop? Why is it using move and collide and move and slide simultaneously? Why does move and slide do the shift? What does it mean by shifting up by pixel steps and then shift back down? Why do we need to shift with pixel steps? How is gravity angle and up direction utilized in this class?

These are some questions I have, but I guess the main one I have is how move and collide, move and slide, and test move are utilized to achieve seamless 360 degree movement along any slope. Is there a resource or guide I am unaware of that explains a lot of the underlying theory behind it? I am aware of the sonic physics guide on sonic retro, but I'm unsure how it's being used here if it's being referenced from to get all the physics correct. Is there maybe some other resource or guide that explains a lot of the physics and complex collision detection for 360 degree momentum based movement games like sonic?

Please let me know about these resources as I'd like to learn more about the physics and programming behind these kind of games. Just like you all, I am a big fan of sonic movement and what you all made here with this framework is just really cool and I just want to know how its done in Godot since there aren't many resources out there that go as in depth as this project. I'd really appreciate the help so again please let me know more about how you all made this work in Godot.

**Comment by DimensionWarped** [comment]
I believe the idea with cutting a given move into steps is that you can be running along a curve at a high enough speed that if your physics is only being calculated at say 60hz you would need to cut your movement for that frame into discrete portions so that you can do the following:

1. move the player in accordance with their direction of travel
2. Reposition the player so that they are on the ground
3. Calculate a new direction of travel based on the slope of the ground that they are tracking to.

If your speed is especially fast and you go over a steep curve that you want the player to stick to, like this one...

![image](https://github.com/Techokami/SonicWorldsNext/assets/7477617/5d574b85-e876-43ff-a341-5924077be134)

it can be better to segment that move into chunks and take it as if you were doing (let's say) 5 separate smaller movements.

In the normal case, this makes it so that your movement flows more smoothly since you don't have big periods where you were moving relative to the slope at the beginning and instead the average movement for the frame starts to be more like the middle point of where it should have been. On the extreme case, you could slip off because you get calculated as too far from the slope to stick to it when you are done with the step and fly off.

I'm not sure how we are doing it exactly right now... more of a @Renhoex question, but ideally the number of intermediary steps would be based on the player's ground speed. Maybe not though since the performance hit doing those calculations makes (which in fair shouldn't be high since this isn't fundamentally different from what the Megadrive was doing back on an 8 Mhz processor) would be better left consistent per frame.

One property of move and slide worth noting is that if the move would have resulted in the object entering the ground, it will be ejected from the ground without our intervention. The 'slide' part means that it doesn't just stop at exactly the point it contacted at, it continues to push in the vector of the movement with just the component that would have been stopped by the collision. It just makes the collision behave a little more organically.

Truthfully, I haven't dug a great deal deep into the specifics of the code here, but I suspect move_remaining is part of what I was describing above... you cut the movement into manageable steps, apply it one step at a time, and then you finish your loop once there is no longer movement that needs to be processed. I'm not sure though. Again, more of a @Renhoex question.

**Reply by Renhoex** [reply]
I coded the physics to try and stick to godot's physics system, it's been a while since I wrote the code but if memory server, move and collide is only used to help detect objects and move_and_slide is just used by default as the mover.

The real collisions happen with the raycast system, I coded it to try and follow the sonic physics guide on sonic retro where it sticks to a 90 degree snap and uses the closest collision to try and calculate the ground normals.

DimensionsWarped pretty much got it. the movement splits is just me being precautions so that sonic can move ludicrously fast without detaching from the floor, like say he's moving 32 pixels, that means he's going to shift 32 pixels to the right before any slope calculations get done, my system I coded it to split up when you go over 8 pixels, so it goes by 8's until it hits 32 for example so that way sonic's always following the curve until he runs out of movement.

Sorry if this response wasn't detailed I'm in a rush right now
